collections-csharp/UpdateItemInListCSharp/UpdateItemInListCSharp/Library.cs

```diff
@@ -0,0 +1,86 @@
+﻿using UpdateItemInListCSharp.Models;
+
+namespace UpdateItemInListCSharp
+{
+    public class Library
+    {
+        public List<Book> Books { get; set; }
+
+        public Library()
+        {
+            Books =
+                [
+                new("Harry Potter and the Philosopher's Stone", "J.K. Rowling", "978-0439708180", false),
+                new("Harry Potter and the Chamber of Secrets", "J.K. Rowling", "978-0439064873", false),
+                new("Harry Potter and the Goblet of Fire", "J.K. Rowling", "978-0439139601", false),
+                new("Harry Potter and the Order of the Phoenix", "J.K. Rowling", "978-0439358071", false),
+                new("Harry Potter and the Half-Blood Prince", "J.K. Rowling", "978-0439785969", false)
+            ];
+        }
+
+        public Book AddBook(string title, string author, string isbn, bool isCheckedOut)
+        {
+            var book = new Book(title, author, isbn, isCheckedOut);
+            Books.Add(book);
+            return book;
+        }
+
+        public void CheckoutBookUsingFind(string isbn)
+        {
+            var bookToCheckout = Books.Find(b => b.ISBN == isbn);
+            if (bookToCheckout != null)
+            {
+                bookToCheckout.IsCheckedOut = true;
+            }
+        }
+
+        public void CheckoutBookUsingFindIndex(string isbn)
+        {
+            var index = Books.FindIndex(b => b.ISBN == isbn);
+            if (index != -1)
+            {
+                Books[index].IsCheckedOut = true;
+            }
+        }
+
+        public void CheckoutBookUsingFirstOrDefault(string isbn)
+        {
+            var bookToCheckout = Books.FirstOrDefault(b => b.ISBN == isbn);
+            if (bookToCheckout != null)
+            {
+                bookToCheckout.IsCheckedOut = true;
+            }
+        }
+
+        public void CheckoutBookUsingForeach(string isbn)
+        {
+            foreach (var book in Books)
+            {
+                if (book.ISBN == isbn)
+                {
+                    book.IsCheckedOut = true;
+                }
+            }
+        }
+
+        public void CheckoutBookUsingIndexOf(Book book)
+        {
+            var newBook = new Book(book.Title, book.Author, book.ISBN, book.IsCheckedOut);
```

Comment:
We should not created a newBook object here, he we should just use the book parameter passed in the method. The caller should create a new book object(different one then it used to Add in the list) for searching.

Reply:
Would this be enough? 
library.CheckoutBookUsingIndexOf(library.Books[0]);

Reply:
That is again not the correct way, in that way you will again end up using the object reference present in the Books list in library class. You should simply create the new Book object in the caller with the Title, Author and ISBN property set to the value of one of the existing object present in Books list and pass it to CheckoutBookUsingIndexOf() method.

Instead of the below line in CheckoutBookUsingIndexOf() method:
var newBook = new Book(book.Title, book.Author, book.ISBN, book.IsCheckedOut);

I suggested to have this line, but in the caller which will created new object reference, and then passed that object as parameter to CheckoutBookUsingIndexOf() method.